### ISSUE TITLE
block-closing-brace-newline-* and at-rule-empty-line-* for some at-rule

### ISSUE THREAD

**Issue by evilebottnawi** [stylelint/stylelint#550]
```scss
@mixin line-on-sides($count-lines: 1, $width: 145px, $color: #fff, $margin: 20px) {
    display: inline-block;
    position: relative;
    vertical-align: middle;
    text-align: center;
    line-height: .5;

    &:before,
    &:after {
        @if $count-lines == 1 {
            border-top: 1px solid $color;
            height: 3px;
        } @else if $count-lines == 2 {
            border-top: 1px solid $color;
            border-bottom: 1px solid $color;
        } @else {
            @error "$count-lines may be 1 or 2, was #{$count-lines}.";
        }

        position: absolute;
        top: 50%;
        width: $width;
        height: $count-lines * 2 + 1px;
        content: "";
    }

    &:before {
        right: 100%;
        margin-right: $margin;
    }

    &:after {
        left: 100%;
        margin-left: $margin;
    }
}
```

Errors:
```bash
13:10	⚠  Expected newline after "}" (block-closing-brace-newline-after) [stylelint]
13:11	⚠  Expected empty line before at-rule (at-rule-empty-line-before) [stylelint]
16:10	⚠  Expected newline after "}" (block-closing-brace-newline-after) [stylelint]
16:11	⚠  Expected empty line before at-rule (at-rule-empty-line-before) [stylelint]
```

I think we need discussion about `block-closing-brace-newline-after`, `block-closing-brace-newline-before`, `at-rule-empty-line-before`, `at-rule-empty-line-after` for control structure.
What do you thinks?

**Comment by davidtheclark** [comment]
@evilebottnawi : You're raising all these SCSS-specific bugs! I wouldn't mind a PR adding an `ignore: ["@else"]` option to these rules. Anybody else have an opinion?

**Comment by evilebottnawi** [comment]
@davidtheclark i think another parsers also have specific at-rule. I think add option ignore solve the problem. But best solution i think (example for block-closing-brace-newline-*):
```json
"block-closing-brace-newline-before": {
    2,
    [
        {
            "type": ["rule"],
            "value" : "always"
        },
        {
            "type": ["at-rule"],
            "name": ["media", "import"],
            "value" : "always"
        },
        {
            "type": ["at-rule"],
            "name": ["if", "else"],
            "value" : "never"
        }
    ]
}
```
More powerfull. I could manage any design at-rule and lint it. Also save short version rules. What do you think about more powerfull rule? @davidtheclark 
About `at-rule-empty-line` try to do roughly the same configuration.

**Comment by jeddy3** [comment]
> i think another parsers also have specific at-rule.

@davidtheclark I was making a start on #543 when I started thinking a little more supporting non-standard parsers and our goals for this project.

It seems our intent at the beginning was to create a linter for present and future CSS syntax. Maxine even raised the possibility of the custom parsers coming to postcss [way back in July](https://github.com/stylelint/stylelint/issues/266#issuecomment-120596014), and we decided then to focus our efforts on standard CSS syntax (because of the amount of work involved and because, I believe at the time, all three of us no longer used sass or less). 

We also anticipated users introducing non-standard CSS syntax through postcss plugins, and so we drafted the first version of the [CSS Processors doc](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/css-processors.md#css-processors) (putting the user in charge of balancing plugin features with their desire to lint code) and we agreed to discuss non-standard syntax introduced by plugins on an case-by-case-basis, which we did. 

Then, not so long ago, the `postcss-scss` custom parser appeared and we added it to the linter without much discussion. I suspect neither of us remembered that conversation from back in July, and the warning it could open a can of worms.

This is probably a good time, with the reporting of these scss-specific issues, to discuss what opening that can means for us. 

The linter landscape has changed some what since we started this project. There are now node-based linters for [sass](https://github.com/sasstools/sass-lint) and [less](https://github.com/lesshint/lesshint). Each has plenty of rules specific to sass and less features/syntax respectively. Just like stylelint has rules specific to future css syntax:

* `function-calc-no-unspaced-operator` - calc
* `function-space-after` - the initial use case was for the new `color()` function.
* `custom-property-no-outside-root` - custom properties
* `custom-property-pattern` - custom properties
* `root-no-standard-properties` - custom properties
* `media-feature-range-operator-space-after` - range operator
* `custom-media-pattern` - custom media

So, the question we need to ask ourselves is... what do we want this tool to be? And, more importantly, what do we want to put our effort into? For me, I'm not really interested in adding support for non-standard syntax (for example, addressing the `// comment` bugs) as it's not something I'm ever likely to use. You might feel differently though. I understand that the division of postcss and scss isn't as clear cut as that, and that postcss can be used as a companion to scss. So, the question is... do we want stylelint to be a linter for postcss parsers, or do we want it to be narrowly-focused on standard CSS? There's no right answer here and I'm looking for your thoughts on the matter.

I think there's a user base out there for simple CSS (CSS modules, [CSS custom properties](http://philipwalton.com/articles/why-im-excited-about-native-css-variables/), Web components, [calc](http://caniuse.com/#feat=calc) etc...). I think what we've have built, and what we have now, is a linter for these people. Where, if anywhere, do we want to take it next?

**Comment by davidtheclark** [comment]
@jeddy3 Thanks for laying that out. Here's what I'm thinking right now about the standard vs non-standard question:

I do not want to build any rules that target non-standard syntax specifically. So, for example, I don't want any rules that specifically target names for `$` variables, or space for non-standard math (e.g. `margin: $thing + 2px`), or nesting. I agree that those should be handled by language-specific linters like those you linked to or whenever possible stylelint plugins. I think we're in agreement there.

I do think that what's here in stylelint is a *ton* of good work, and I want to make that work available to as many users as possible. Additionally, in my ideal world these Node-based Sass and Less linters would focus on Sass/Less specific issues, while actually suggesting stylelint for the other stuff — instead of writing their own rules about everything that we've already tackled. To that end, I would like non-standard syntax users to be able to run their (non-processed) code through stylelint whenever it's reasonable. If a single extra conditional, or something else simple, allows SCSS users to take advantage of some rule, I am willing to make that concession. If the concession turns out to be significant, though, I would err on the side of simplicity and expect SCSS users not to use the rule, or to use the ignore comments.

Maybe one way to summarize is this: I would really like stylelint to be able to *not choke* on anything that PostCSS itself doesn't choke on. Typically that means that it doesn't really have to understand what's going on in potentially crazy syntax.

In situations like this issue, I'd like to think about the problem, maybe experiment a little, and judge from there whether it's practical to accommodate the request.

So to sum up ... In terms of strict principles, I'd say I'd like to *not* write rules that specifically target non-standard syntax, but *try* to make the linter *work* with non-standard syntax that PostCSS understands, whenever reasonable.

**Comment by evilebottnawi** [comment]
@davidtheclark @jeddy3 I think that the problem can be solved by specific syntax's by adding specific rules for these syntax's (eg Sass, less, and others).  Stylelint remain independent for syntax and his rules should be based only on the AST tree, forming postcss. In the future it may appear even more new and innovative styles description different from the specification css. And there's no getting around it. Therefore, I see styling as linter giving the opportunity to lint their styles, based on the generated AST. It should not contain the rules for the other syntax's. But it should allow a complete lint your code being generated on the basis of AST tree. My proposed rule above generally do not aim to introduce specific rules for the scss. It is aimed at a more detailed inspection of the formation of the brackets for certain ```javascript node.type ``` generated by postcss. Is not it? Sorry for my bad english

**Comment by davidtheclark** [comment]
> Stylelint remain independent for syntax and his rules should be based only on the AST tree, forming postcss

That's a good way of putting it. I'd like stylelint to be able to lint the AST produced by PostCSS, whatever the input.

**Comment by jeddy3** [comment]
Thanks for the well-thought-out replies. I think we're all on the same page here, and I agree that @evilebottnawi has expressed, quite succinctly, what that page is.

> I do think that what's here in stylelint is a ton of good work, and I want to make that work available to as many users as possible. 

It's great to understand the motivation behind the decision making, especially when it comes to discussing the merits of each concession.

> Additionally, in my ideal world these Node-based Sass and Less linters would focus on Sass/Less specific issues, while actually suggesting stylelint for the other stuff — instead of writing their own rules about everything that we've already tackled.

I agree. Shall we reach out to them?

> In situations like this issue, I'd like to think about the problem, maybe experiment a little, and judge from there whether it's practical to accommodate the request.

That makes sense. Unfortunately, for the reason I communicated privately to you, I won't be able to get involved in this process as much as I'd like. The number of contributors has grown though, and so hopefully you'll have someone (like @evilebottnawi) to bounce ideas off.

> @evilebottnawi: My proposed rule above generally do not aim to introduce specific rules for the scss. It is aimed at a more detailed inspection of the formation of the brackets for certain javascript node.type generated by postcss. Is not it?

Yep, it is :)

**Comment by evilebottnawi** [comment]
@jeddy3 @davidtheclark About Node-based Sass and Less linters and another. Fortunately or unfortunately, most of them are dead and not being developed further, or too slow. By the way, this is one of the reasons why I made the choice to stylelint. So I think in the near future the number of people using specific syntaxes increase here. As I said earlier is probably correct to develop specific rules for them throw plugins. Stylelint itself to remain clean of all specific rules for specific syntax. Letting those who write on specific syntax and develop specific rules.

**Comment by davidtheclark** [comment]
On the issue at hand ... @evilebottnawi I think that the syntax you propose for options is more verbose than it needs to be. What if we use `ignore` and `except`, as in other rules, where the values in those arrays are at-rule names?

Your example would become:

```json
"block-closing-brace-newline-before": { 2, { "except": ["else"] } }
```

**Comment by evilebottnawi** [comment]
@davidtheclark I think at this point it would be enough if they appear still problems with this rule will be consider more detailed customization rule.

**Comment by davidtheclark** [comment]
Ok, I am going to open a new issue for that, so that if somebody wants to submit a PR they have a clean issue to work from.
